### ISSUE TITLE
eigen in non-diagonal matrices

### ISSUE THREAD

**Issue by ronisbr** [JuliaArrays/StaticArrays.jl#661]
Hi!

I am getting this error when trying to use `eigen` on non-diagonal matrices:

```julia
julia> Imc = @SMatrix([446. 100. 0.; 0. 1149. 100.; 0. 0. 886])
3×3 SArray{Tuple{3,3},Float64,2,9} with indices SOneTo(3)×SOneTo(3):
 446.0   100.0    0.0
   0.0  1149.0  100.0
   0.0     0.0  886.0

julia> eigen(Imc)
ERROR: Only hermitian matrices are diagonalizable by *StaticArrays*. Non-Hermitian matrices should be converted to `Array` first.
Stacktrace:
 [1] error(::String) at ./error.jl:33
 [2] _eig at /Users/ronan.arraes/.julia/packages/StaticArrays/3KEjZ/src/eigen.jl:120 [inlined]
 [3] #eigen#492 at /Users/ronan.arraes/.julia/packages/StaticArrays/3KEjZ/src/eigen.jl:386 [inlined]
 [4] eigen(::SArray{Tuple{3,3},Float64,2,9}) at /Users/ronan.arraes/.julia/packages/StaticArrays/3KEjZ/src/eigen.jl:386
 [5] top-level scope at REPL[43]:1
```

It works with diagonal matrices:

```julia
julia> Imc = @SMatrix([446. 0. 0.; 0. 1149. 0.; 0. 0. 886])
3×3 SArray{Tuple{3,3},Float64,2,9} with indices SOneTo(3)×SOneTo(3):
 446.0     0.0    0.0
   0.0  1149.0    0.0
   0.0     0.0  886.0

julia> eigen(Imc)
Eigen{Float64,Float64,SArray{Tuple{3,3},Float64,2,9},SArray{Tuple{3},Float64,1,3}}
eigenvalues:
3-element SArray{Tuple{3},Float64,1,3} with indices SOneTo(3):
  446.0
  886.0
 1149.0
eigenvectors:
3×3 SArray{Tuple{3,3},Float64,2,9} with indices SOneTo(3)×SOneTo(3):
 1.0  0.0  0.0
 0.0  0.0  1.0
 0.0  1.0  0.0
```

**Comment by c42f** [comment]
I imagine it's possible to fix this by tweaking dispatch so that we just call into LinearAlgebra for non-hermitian matrices. However, that would have type stability consequences for existing users (the resulting matrices being real or complex depending on their value). So we'd have to think what to do about that.

**Comment by CarpeNecopinum** [comment]
My suggestion would be to add an element type parameter to `eigen` that defaults to `eltype` of the input matrix. In the case of a non-hermitian matrix, we call `Base.eigen` and check if the result can be fit into static arrays with the given element type, otherwise we throw an exception.

That way, we keep any code that is type stable today type stable and working in the new variant, while adding a way to handle non-hermitian matrices.

I can cook up a PR if desired.

**Comment by c42f** [comment]
Fixed by #683
